### PR TITLE
OpenStreetMap handler Integration

### PR DESCRIPTION
**Pull request by singh1203**
## Description

OpenStreetMap API Integration 

**Fixes** #5550

## Type of change

(Please delete options that are not relevant)

- [X] ⚡ New feature (non-breaking change which adds functionality)

### What is the solution?

I have created a class for the OpenStreetMap handler and for the Nodes, Ways, and Relation API Table. Also, I have implemented the Overpass API method for API integration instead of the main API method, Since overpass API follows box modal instead of XML.

## Checklist:

- [X] My code follows the style guidelines(PEP 8) of MindsDB.
- [X] I have commented on my code, particularly in hard-to-understand areas.
- [X] I have updated the documentation, or created issues to update them.

### CODE REVIEW

**Comment by singh1203** [comment]
I have read the CLA Document and I hereby sign the CLA

**Comment by mindsdbadmin** [comment]
**CLA Assistant Lite bot:** <br/>Thank you for your submission, we really appreciate it. Like many open-source projects, we ask that you sign our [Contributor License Agreement](https://github.com/mindsdb/mindsdb_native/blob/stable/assets/contributions-agreement/individual-contributor.md) before we can accept your contribution. You can sign the CLA by just posting a Pull Request Comment same as the below format.<br/>
   - - -
   I have read the CLA Document and I hereby sign the CLA
   - - -
   <sub>You can retrigger this bot by commenting **recheck** in this Pull Request</sub>

**Comment by singh1203** [comment]
I have read the CLA Document and I hereby sign the CLA

File: mindsdb/integrations/handlers/openstreetmap_handler/openstreetmap_handler.py [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+    
+    def check_connection(self) -> StatusResponse:
+        """Check connection to the handler.
+
+        Returns:
+            HandlerStatusResponse
+        """
+        try:
+            self.connect()
+            return StatusResponse(status=True)
```
Comment (MinuraPunchihewa) [inline mindsdb/integrations/handlers/openstreetmap_handler/openstreetmap_handler.py:67]: @singh1203 In instantiating objects of the `HandlerStatusResponse` and `HandlerResponse`, you seem to be passing in a parameter called `status`. However, it does not look like either of these classes accept such a parameter. I am not able to execute `CREATE DATABASE` statements. as a result.

**Review by MinuraPunchihewa — CHANGES_REQUESTED** [review]

File: mindsdb/integrations/handlers/openstreetmap_handler/README.md [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+WITH
+engine='openstreetmap',
+parameters={
+    "area": "New York City",
+    "timeout": 1000
+};
+~~~~
+
+```sql 
+SELECT * FROM nodes LIMIT 10;
```
[the remark this answers is not in the document]
Comment (singh1203) [reply mindsdb/integrations/handlers/openstreetmap_handler/README.md:76]: Sure, I'll be implementing in that way.

File: mindsdb/integrations/handlers/openstreetmap_handler/openstreetmap_handler.py [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+            Response: The response from the query.
+        """
+        try:
+            self.connect()
+            result = self.connection.query(query)
+            return Response(status=True, data=result)
+        except Exception as e:
+            return Response(status=False, message=str(e))
+        
+    def call_openstreetmap_api(self, method_name:str = None, params:dict = None) -> pd.DataFrame:
```
[the remark this answers is not in the document]
Comment (singh1203) [reply mindsdb/integrations/handlers/openstreetmap_handler/openstreetmap_handler.py:87]: Sorry for that @MinuraPunchihewa, It was just for a rough try and forget to remove it. Also I will be making PR  as a draft PR for as if for now and Resolving all of the suggested reviews.

[the remark this answers is not in the document]
Comment (MinuraPunchihewa) [reply mindsdb/integrations/handlers/openstreetmap_handler/openstreetmap_handler.py:87]: Thanks @singh1203, let us know if we can help in any way.

**Review by MinuraPunchihewa — CHANGES_REQUESTED** [review]

**Comment by ZoranPandovski** [comment]
@singh1203 Is this ready?

**Comment by singh1203** [comment]
> @singh1203 Is this ready?

@ZoranPandovski Almost, I am seeking the help of @MinuraPunchihewa he'll review from his end and will advise me for the rest.

**Comment by MinuraPunchihewa** [comment]
I will review this today @ZoranPandovski @singh1203.

**Comment by MinuraPunchihewa** [comment]
@ZoranPandovski There seems to be an issue with running SELECT statements after creating a connection. @singh1203 will try to figure out what the issue is.

**Comment by MinuraPunchihewa** [comment]
Hey @singh1203,
From what I have gathered, the reason that your `SELECT` statements seem to be hanging is because the query that your `Table` classes are executing within the `select()` are very general. For instance, the `select()` method of the  `OpenStreetMapNodeTable` is attempting to extract all of the nodes, which I suspect can be quite intensive. Using the Overpass Query Language, you can apply filters for what you want to extract,
https://wiki.openstreetmap.org/wiki/Overpass_API/Overpass_QL

I have written a function called `execute_osm_node_query()` that generates and executes these queries (for nodes) based on the conditions that are passed.

I have expanded on the code that is used to parse the query as well. Since this is not a typical type of source, it does not make a lot of sense to use the `SELECTQueryParser`.

Note that the default number of items (nodes) that will be extracted has been set to 20. Essentially, if you run a `SELECT * FROM osm_datasource.nodes` statement, only 20 records will be returned.

To complete this handler, I recommend that you do the following,
1. Please take a look at the updated the code for `OpenStreetMapNodeTable`, especially the `execute_osm_node_query()`. Although it seems to be quite robust, I am not a 100% sure that I have incorporated all of the intricacies of the Overpass Query Language.
2. A similar logic will need to be added to the other tables as well. These classes should include the logic to execute queries specific to the type of data they will be interacting with.

If you think that you want to tackle number 2 above later on, that is fine too. We can support nodes for now, as long as we address number 1 and make sure it is robust.

**Review by ZoranPandovski — CHANGES_REQUESTED** [review]
Can you please solve the pep8 warnings before we merge this https://github.com/mindsdb/mindsdb/actions/runs/6472810891/job/17574180182?pr=6740#step:5:12

**Comment by singh1203** [comment]
> Can you please solve the pep8 warnings before we merge this https://github.com/mindsdb/mindsdb/actions/runs/6472810891/job/17574180182?pr=6740#step:5:12

I appreciate you alerting me to the PEP8 warnings, @ZoranPandovski, and I believe we may move forward at this time.

**Review by ZoranPandovski — CHANGES_REQUESTED** [review]
Awesome, before merging this as requested in the PR checklist can you share few screenshots or a short video of working integrations and add initial unit tests please? Thanks

**Comment by singh1203** [comment]
Screenshot from 2023-11-02 22-31-57
Screenshot from 2023-11-02 23-07-23
Screenshot from 2023-11-02 23-08-56

**Comment by singh1203** [comment]
@ZoranPandovski Once again, thank you. I tried my best to fulfill your request and have supplied a few screenshots above for your reference.

**Review by ZoranPandovski — APPROVED** [review]

**Pull request merged**